### PR TITLE
VMware: vmware_tools connection plugin

### PR DESCRIPTION
**Pull request by jamescassell**
##### SUMMARY

I've updated the referenced PR 

Fixes: #47072

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

vmware_tools

##### ADDITIONAL INFORMATION

> > At one point there was a wiki page attempting to document and standardize these going forward, but I'm not sure where that is anymore. I think the only things we all agree on are `ansible_user` and `ansible_password` defaulting for the vm username and password. smile
> 
> Do you mean this - https://github.com/ansible/community/wiki/VMware:-standardization
> 

I've integrated the changes from TLS standardization and connection_var standardization.  I also added support for the VI_* env vars standardized by vmware sdk.

> > I could use some help getting started or some examples of what this needs to look like. Is there a good reference for this in the existing `ansible` codebase?
> 
> I am sure that we can do integration test as govcsim does not emulate actual OS. So unit test will be more than sufficient for this change. You may want to follow existing connection plugin testcases - https://github.com/ansible/ansible/tree/devel/test/units/plugins/connection
> 

I'm punting this for now.  I'd want to do an integration test, but that requires a full VMware setup.

> > > 1. End-user documentation
> > 
> > 
> > I assumed that meant something like the [winrm connection page](https://docs.ansible.com/ansible/latest/plugins/connection/winrm.html), but I thought that would be autogenerated from the `DOCUMENTATION` block. Either way, do we need more?
> 
> Yes, we can add documentation here = https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/vmware/ as how-to-use connection plugin.
> 

I added an EXAMPLES block.

> > I'm not sure how to go about doing that here, it's potentially something different altogether, see the discussion in #43428.
> > I believe the plugin works, as (manually) tested by myself and at least a couple of others in this thread. I think we just need to decide what it needs for the 2.8 release. I'm willing to finish it up, but I need use some clearer definitions of what _finished_ looks like.
> 
> Not sure about this. But I think if we can get initial version in tree, we can always improvise things. Its better to have something than nothing. :)

This PR intends to show what 'finished' looks like.

```paste below

```

### CODE REVIEW

**Comment by jamescassell** [comment]
@Akasurde @dericcrago @pdellaert @Im0 @tchernomax @dagwieers @jborean93 @nitzmahone @verdel PTAL.  This is a cleanup of #47072 addressing most of the feedback.

**Comment by Akasurde** [comment]
@pdellaert @Im0 @ckotte @jeking3 @Tomorrow9 @alongchamps @adarobin @goneri  Could you please review this ?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
    short_description: Execute tasks inside a VM via VMware Tools
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
      - Use VMware tools to run tasks in, or put/fetch files to VM running in VMware infrastructure.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
      - In case of Windows VMs, set C(ansible_shell_type) to C(powershell).
```
I am OK with current description as well.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
      - Does not work with 'become'.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
      - pyvmomi (Python library)
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
      - requests (Python library)
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - FQDN or IP Address for the connection (vCenter or ESXi Host).
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:56]:
```suggestion
          - name: VMWARE_HOST
```
This will match [existing environment variable](https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L57) used all VMware modules.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:70]:
```suggestion
          - name: VMWARE_USER
```
Same as VMWARE_HOST - https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L63

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:79]:
```suggestion
          - name: VMWARE_PASSWORD
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:101]:
```suggestion
        env:
          - name: VMWARE_VALIDATE_CERTS
        vars:
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - VM path absolute to the connection.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - "vCenter Example: C(Datacenter/vm/Discovered virtual machine/testVM)."
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - "ESXi Host Example: C(ha-datacenter/vm/testVM)."
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - Must include VM name, appended to 'folder' as would be passed to M(vmware_guest).
          - Needs to include C(vm) between the Datacenter and the rest of the VM path.
          - Datacenter default value for ESXi server is C(ha-datacenter).
          - C(vm) is not visible in vSphere Web Client but necessary for VMware API to work.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:126]:
```suggestion
          - Username in guest operating system.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - Password for the user in guest operating system.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
          - Time in seconds to sleep between execution of command.
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:175]:
```suggestion
ansible_vmware_validate_certs: False # default is True
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
import traceback

REQUESTS_IMP_ERR = None
try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    REQUESTS_IMP_ERR = traceback.format_exc()
    HAS_REQUESTS = False
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:152]: Also add parameter `VMWARE_PORT`. Sometime vCenter is located on custom port rather than `443`. 

https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L85

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:336]: Add same import error block for requests.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py:450]: how about - 

https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/modules/cloud/vmware/vmware_vm_shell.py#L303

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@jamescassell Thanks for working on this. Let me know if you need any help with this.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
- name: Test VMware Tools Connection Plugin for Linux
```

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]:
```suggestion
- name: Test VMware Tools Connection Plugin for Windows
```

**Comment by Akasurde** [comment]
@jamescassell I added changes here - https://github.com/jamescassell/ansible/pull/1 feel free to modify.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py:450]: Not blocker for now.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:56]: I'd like to support both if you don't mind.  Is VMWARE_HOST defined anywhere outside of ansible?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py]: thanks for the extra detail!

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:175]: why? (no big deal, just curious)

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:152]: Yeah, would be good to add... do you immediately know how it would be passed?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py:56]: `VMWARE_HOST` is environment variable since inception of VMware_* modules. Even, I don't mind supporting both.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py]: No problem.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py:175]: Nothing specific, wanted to match other module style.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py:152]: It is available in PR. Added on https://github.com/ansible/ansible/pull/55059/files#diff-2c15779308a54e8ac33b5eba806f6e5bR84

Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:152]: nevermind, I see that you added it

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [inline lib/ansible/plugins/connection/vmware_tools.py:284]: I assume it's fine to use get_option here, too?

Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:284]: Seems to have been fine... I made the change and tested that things still work.

Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:284]: okay, I now see it was an '@ property' so the get_option was elsewhere

**Comment by jamescassell** [comment]
@Akasurde I added your changes, re-added the VI_* env vars, and then squashed to one commit per contibutor so ansibot will be happy to merge it once approved.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py]: dont use play_context, add 'executable' option to docs and `self.get_option('executable')`

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:22]: i would avoid using requests, introduced dep that is not needed, as functions in urls.py  should cover what this plugin uses.

**Comment by jamescassell** [comment]
@Akasurde I've integrated the last two of your suggestions.  (Shippable seems broken, however... which seems unrelated)

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:105]: how are these warnings and not failures? is this controlling behaviour if `validate_certs` is False? would we not just ignore warnings then?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:444]: should these be AnsibleConnection errors instead?

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py]: What should be the default value?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:105]: It was added by the original author.  @dericcrago is it okay to remove this option and have it default to 'true'?

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:444]: I don't know, but I can change it if it should be.

**Comment by Akasurde** [comment]
@jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py]: I think @bcoca meant something like this - https://github.com/ansible/ansible/blob/166a33ef4d7272ed83c0b365545ccaccc50dd6d5/lib/ansible/plugins/connection/lxd.py#L23

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py:105]: I think we can set it to True by default so that we can just ignore warnings as user set `validate_certs` to `False`

**Comment by jamescassell** [comment]
> @jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.

Looks like it boils down to:
- adding `self.get_option('executable')` preferably with a `ansible_vmware_tools_executable` var for configuration
- dropping `silence_tls_warnings`
- AnsibleError -> AnsibleConnection error
- swap python requests for Request from module_utils/urls.py

I can likely get the first 3, but I need to know what the default value should be for `executable`

I'm not sure if I can just drop/replace `requests.get` with `Requests.get` ?

I won't be able to test any changes I make until 7 hours from now.  (I have, of course, tested the PR as-is.)

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (Akasurde) [inline lib/ansible/plugins/connection/vmware_tools.py]: @jamescassell I think we should remove this as makedirs_safe is not safe. We should not create any directory as user should ensure creation of directory rather than creating them silently.

Comment (abadger) [reply lib/ansible/plugins/connection/vmware_tools.py]: s/user/caller/  (although user creation really is the only safe method if the file is in a world-writable directory).

Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py]: Yup. Totally agreed.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (goneri) [inline lib/ansible/plugins/connection/vmware_tools.py:156]: Just a suggestion. We can rely on the GuestID here to set the default value: If the GuestId starts with "win" -> PowerShell, otherwise, /bin/sh.

Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:156]: I think this is handled by setting ansible_shell_type=powershell

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:22]: I agree in principle that the built-in methods should be sufficient, as it's doing one `requests.get` and one `requests.put` ...but I don't know how to translate between the two...

On further inspection, it looks like the built-in urls.py doesn't support chunk_size and piecemeal fetching of remote files, so I'm inclined to just leave the requests dep as-is.

This is only a dep for the controller, not the managed nodes, and all other vmware components in ansible also require python requests

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py]: Might need better wording, but ignoring for now.  I accepted your change on this item.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py]: Done.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:105]: Now defaults to `true`

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:444]: Done.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py]: Done.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:156]: The plugin already hardcodes it to "cmd.exe" if the guestfamily is windows.

**Comment by jamescassell** [comment]
> > @jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.
> 
> Looks like it boils down to:
> 
> * adding `self.get_option('executable')` preferably with a `ansible_vmware_tools_executable` var for configuration

Done.

> * dropping `silence_tls_warnings`

Changed default to `true` as some thing it's useful.

> * AnsibleError -> AnsibleConnection error

Done.

> * swap python requests for Request from module_utils/urls.py
> 

Will not do, as urls.py doesn't support chunked downloads and the rest of vmware in ansible also requires requests.



Otherwise, I think I've addressed all feedback.

ready_for_review

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (felixfontein) [inline lib/ansible/plugins/connection/vmware_tools.py]: This isn't true anymore.

Comment (Akasurde) [reply lib/ansible/plugins/connection/vmware_tools.py]: Good catch.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (gundalow) [reply lib/ansible/plugins/connection/vmware_tools.py:22]: Can you please raise an issue for this so we can get the built in updated

**Review by goneri — APPROVED** [review]

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [reply lib/ansible/plugins/connection/vmware_tools.py:22]: if its already a pvimomi dependency its ok to use, but then you really dont need to manage it this way, just import it in same block you do pyvimomi and only deal with missing pyvimomi

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:164]: missing ini/env var entries (to use global default)

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:42]: i would mention this plugin is Windows specific

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (bcoca) [inline lib/ansible/plugins/connection/vmware_tools.py:277]: these settings make this a 'windows only' connection , but examples above suggest otherwise, which is it?

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:42]: Plugin works for both linux and windows.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:164]: Added.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:277]: It supports both.  If SHELL_FAMILY=powershell, it does this special magic.  If not, it "just works" for Linux otherwise.

File: lib/ansible/plugins/connection/vmware_tools.py
Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:22]: I'll raise an issue.  On my RHEL7 system, requests doesn't seem to be a direct dep of pyvmomi, but on my Fedora 29 system, python3-requests is a direct dep of python3-pyvmomi.

Comment (jamescassell) [reply lib/ansible/plugins/connection/vmware_tools.py:22]: Issue opened: https://github.com/ansible/ansible/issues/55162

**Review by Akasurde — APPROVED** [review]

**Comment by jamescassell** [comment]
Thanks @Akasurde for the fixes!  I've also addressed the fix suggested by @bcoca of also enabling ini and env for the executable setting.

Let me know if I should re-squash to one commit per author.

ready_for_review

**Comment by jamescassell** [comment]
@Akasurde thanks for all your help on this and for merging! @dericcrago thanks for writing it to begin with!

**Pull request merged**
